### PR TITLE
Add comments to BoolOps implementation.

### PR DESCRIPTION
**Pull request by andriyDev**
- [X] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [ ] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---
This part of the code relies on a lot of assumptions that are not obviously until you really dig into the code, and have a lot of components that are not really clear (what's the difference between a `Ring` and a `Snake`?).

I've taken a stab at documenting these things and adding a few more inline comments to hopefully make it more clear to future readers.

There are a couple parts I still don't really understand, which I've marked as TODOs. Hopefully with a review, someone else might have more insight and we can document this even better. I will also likely expand the scope of comments as I make more progress (so far this is just `assembly.rs`, but I may try to capture more of the BoolOps).

### CODE REVIEW

**Comment by rmanoka** [comment]
Thanks @andriyDev ;  the poor docs in this impl. is primarily my fault.  We built this impl. incrementally, trying to fix up the floating point precision issues.  I somewhat gave-up midway, as I couldn't figure how to provably fix the Martinez-Rueda family of algorithms to work with f64/f32.

That said, the Assembly module, iirc, does not have floating point precision issues as it doesn't actually need to break any segments.  In fact, I think, we can even get this module to work without using `GeoFloat` at all (only `GeoNum`).

I'll do my best to review this in the next week or two.

**Comment by andriyDev** [comment]
Hey no blame! The only way to make the code better is to work together :)

As far floating point issues, I had been working on my own Martinez-Rueda implementation and had also been struggling heavily with floating point issues myself. I've got an idea for why those are happening so I'll create a separate issue.

**Comment by andriyDev** [comment]
I rebased onto main, and I added more comments for the `bool_ops` folder.

There are many changes we could make to simplify the code, but I think just documenting it is the first step.

In addition, this isn't the whole story - the main thing that's missing now is documenting `CrossingsIter` which mostly means going through `Sweep` with a fine-toothed comb.

File: geo/src/algorithm/bool_ops/op.rs [hunk 1]
```diff
@@ -173,15 +209,22 @@ impl<T: Float, S: Spec<T>> Proc<T, S> {
...
 }
 
+/// An edge of a shape.
 #[derive(Clone)]
 struct Edge<T: Float, S: Spec<T>> {
+    /// The geometry of the edge.
     geom: LineOrPoint<T>,
+    /// The index of the shape this edge belongs to.
     idx: usize,
+    /// The region of this edge when the piece is going in the same direction as
```
Comment (andriyDev) [inline geo/src/algorithm/bool_ops/op.rs:219]: I'm pretty sure this is wrong. I can sorta understand the intent, but it doesn't make sense to me exactly what the check below means.

**Comment by frewsxcv** [comment]
@andriyDev This is still marked as draft. Is that intentional or is this ready for review?

**Comment by andriyDev** [comment]
I'm happy to call this ready for review. More comments can always be another PR :)

File: geo/src/algorithm/bool_ops/assembly.rs [hunk 2]
```diff
@@ -85,14 +98,15 @@ impl<T: GeoFloat> RegionAssembly<T> {
                         parent_snake_idx = Some(
                             iter.prev_active(d)
                                 .map(|(_, seg)| seg.snake_idx.get())
+                                // TODO
```
Comment (frewsxcv) [inline geo/src/algorithm/bool_ops/assembly.rs:101]: Can you remove these TODOs?

Comment (andriyDev) [reply geo/src/algorithm/bool_ops/assembly.rs:101]: I can, but these TODOs really need an explanation. They make no sense to me. I tried to remove these defaults as well and tests started to fail, so clearly they are necessary but I don't have a logical reason *how*.

Comment (frewsxcv) [reply geo/src/algorithm/bool_ops/assembly.rs:101]: @rmanoka Do you recall why this `.unwrap_or(0)` is here?

Comment (rmanoka) [reply geo/src/algorithm/bool_ops/assembly.rs:101]: I don't think the number here matters.  If the expression is not `Some()`, then the "snake" will complete to an exterior ring, which doesn't need a parent idx.  I tried verifying by setting random numbers inside `.unwrap_or()`, and tests seem to pass.

File: geo/src/algorithm/bool_ops/op.rs (hunk 1 above)
Comment (urschrei) [reply geo/src/algorithm/bool_ops/op.rs:219]: @rmanoka Can you shed any light on this?

File: geo/src/algorithm/bool_ops/assembly.rs (hunk 2 above)
Comment (urschrei) [reply geo/src/algorithm/bool_ops/assembly.rs:101]: @andriyDev Are you OK with this explanation?

**Comment by urschrei** [comment]
I'd really like to see this merged in order to facilitate further investigation into `RegionAssembly`'s tendency to panic (see https://github.com/georust/geo/issues/1104) for a possible avenue to investigate. I think there's only one minor question left to be resolved!

**Comment by andriyDev** [comment]
> I'd really like to see this merged in order to facilitate further investigation into `RegionAssembly`'s tendency to panic (see #1104 for a possible avenue to investigate). I think there's only one minor question left to be resolved!

I would be shocked if this code is "responsible" for the panic. The RegionAssembly is fairly straightforward. What's more likely in my mind is that bad data is coming in from Proc or Spec in boolean ops since those are much more complex IMO. But that's just my speculation haha.

**Comment by urschrei** [comment]
> Regardless we (mostly I) should push this through to make progress on that

Either that or outright replacement with your Martinez-Rueda implementation, but that's all a separate discussion.

**Comment by andriyDev** [comment]
> Either that or outright replacement with your Martinez-Rueda implementation, but that's all a separate discussion.

To be clear, mine is also riddled with panics LMAO. Maybe one day I'll find the "smoking gun".

File: geo/src/algorithm/bool_ops/assembly.rs (hunk 2 above)
Comment (andriyDev) [reply geo/src/algorithm/bool_ops/assembly.rs]: I've added the explanation to the comments. I believe this is ready to go then (unless there are other concerns). :)

**Review by frewsxcv — APPROVED** [review]
@andriyDev thank you so much for adding these comments!

**Pull request merged**